yt/utilities/io_handler.py

```diff
@@ -254,46 +196,31 @@ def _read_particle_selection(
             else:
                 ptf[ftype].append(fname)
                 field_maps[field].append(field)
-        # Now we have our full listing
-
-        # Now we add particle counts across chunks to psize
-        self._count_particles_chunks(psize, chunks, ptf, selector)
+            data[field] = []
 
-        # Now we allocate
-        for field in fields:
-            if field[0] in unions:
-                for pt in unions[field[0]]:
-                    fsize[field] += psize.get(pt, 0)
-            else:
-                fsize[field] += psize.get(field[0], 0)
-        shape: Tuple[int, ...]
-        for field in fields:
-            if field[1] in self._vector_fields:
-                vsize = self._vector_fields[field[1]]  # type:ignore
-                # note: the above line causes a mypy failure due to how we
-                # convert _vector_fields from a tuple to dict in __init__. mypy
-                # is expecting a tuple here. And since _vector_fields is used in
-                # many places, just ignoring for now...
-                shape = (fsize[field], vsize)
-            elif field[1] in self._array_fields:
-                shape = (fsize[field],) + self._array_fields[field[1]]
-            else:
-                shape = (fsize[field],)
-            rv[field] = np.empty(shape, dtype="float64")
-            ind[field] = 0
         # Now we read.
         for field_r, vals in self._read_particle_fields(chunks, ptf, selector):
             # Note that we now need to check the mappings
             for field_f in field_maps[field_r]:
-                my_ind = ind[field_f]
-                # mylog.debug("Filling %s from %s to %s with %s",
-                #    field_f, my_ind, my_ind+vals.shape[0], field_r)
-                rv[field_f][my_ind : my_ind + vals.shape[0], ...] = vals
-                ind[field_f] += vals.shape[0]
-        # Now we need to truncate all our fields, since we allow for
-        # over-estimating.
-        for field_f in ind:
-            rv[field_f] = rv[field_f][: ind[field_f]]
+                data[field_f].append(vals)
+
+        rv: Dict[Tuple[str, str], np.ndarray] = {}  # the return dictionary
+        fields = list(data.keys())
+        for field_f in fields:
+            # We need to ensure the arrays have the right shape if there are no
+            # particles in them.
+            total = sum(_.size for _ in data[field_f])
+            if total > 0:
+                rv[field_f] = np.concatenate(data.pop(field_f), axis=0).astype(
```

Comment:
@chrishavlin maybe you want to refactor this so formatting doesn’t ruin readability?

Reply:
yup! spent a while trying to understand if applying the type conversion to individual chunks before concatenating had any performance differences... to which I can say probably not but maybe??? and so I'm going to keep it here. So will clean it up to be more readable.